jung-visualization/src/main/java/edu/uci/ics/jung/visualization/renderers/BasicRenderer.java

```diff
@@ -23,67 +21,74 @@
 /**
  * The default implementation of the Renderer used by the VisualizationViewer. Default Vertex and
  * Edge Renderers are supplied, or the user may set custom values. The Vertex and Edge renderers are
- * used in the renderVertex and renderEdge methods, which are called in the render loop of the
+ * used in the renderNode and renderEdge methods, which are called in the render loop of the
  * VisualizationViewer.
  *
  * @author Tom Nelson
  */
-public class BasicRenderer<V, E> implements Renderer<V, E> {
+public class BasicRenderer<N, E> implements Renderer<N, E> {
 
   private static final Logger log = LoggerFactory.getLogger(BasicRenderer.class);
-  protected Renderer.Vertex<V, E> vertexRenderer = new BasicVertexRenderer<V, E>();
-  protected Renderer.VertexLabel<V, E> vertexLabelRenderer = new BasicVertexLabelRenderer<V, E>();
-  protected Renderer.Edge<V, E> edgeRenderer = new BasicEdgeRenderer<V, E>();
-  protected Renderer.EdgeLabel<V, E> edgeLabelRenderer = new BasicEdgeLabelRenderer<V, E>();
+  protected Node<N, E> nodeRenderer = new BasicNodeRenderer<N, E>();
+  protected NodeLabel<N, E> nodeLabelRenderer = new BasicNodeLabelRenderer<N, E>();
+  protected Renderer.Edge<N, E> edgeRenderer = new BasicEdgeRenderer<N, E>();
+  protected Renderer.EdgeLabel<N, E> edgeLabelRenderer = new BasicEdgeLabelRenderer<N, E>();
 
   public void render(
-      RenderContext<V, E> renderContext,
-      VisualizationModel<V, E, Point2D> visualizationModel,
-      Spatial<V> spatial) {
-    if (spatial == null) {
+      RenderContext<N, E> renderContext,
+      VisualizationModel<N, E, Point2D> visualizationModel,
+      Spatial<N> nodeSpatial,
+      Spatial<E> edgeSpatial) {
+    if (nodeSpatial == null) {
       render(renderContext, visualizationModel);
       return;
     }
-    Set<V> visibleNodes = null;
+    Set<N> visibleNodes = null;
+    Set<E> visibleEdges = null;
+
     try {
       visibleNodes =
           (Set)
-              spatial.getVisibleNodes(
+              nodeSpatial.getVisibleElements(
```

Comment:
Semantically, the Collection-returning methods of Spatial should probably be Set-returning instead.

Reply:
Well, you've got that typecast in there, which is often (if not always) a sign that you're returning a too-general type.  (It's also a sign that you're depending on the implementation details of the implementation of `Spatial`, which is fragile.)

I generally (heh) prefer to return the most specific semantically appropriate type.  In this case, there should not be any duplicates in the returned collection, so it's semantically a `Set`.

I generally try to design interfaces according to the [Robustness Principle](https://en.wikipedia.org/wiki/Robustness_principle), which is about this kind of question.

Reply:
I will have make visibleNodes an Iterable instead of a Set, since I do not depend on Set semantics and plan only to iterate over it.

Reply:
I don't understand why you wouldn't return a `Set` unless you want to retain the option of internally constructing some other `Iterable` (which seems unlikely considering that in one case, at least, you're already returning a pre-existing `Set` that you'd have to convert).  

I could imagine, for instance, wanting to intersect the set of visible nodes with some other set, which `common.collect.Sets.intersection` can do just fine.  But in any event, you're constructing a `Set` internally (or returning an existing one) so I don't think that there's anything to be lost by reflecting that in the return type, and it gives the user more options for how to use it.

Reply:
good points about the potential for Set operations like intersection